build_runner/test/integration_tests/utils/build_descriptor.dart

```diff
@@ -0,0 +1,305 @@
+
+class TestBuilderDefinition {
+  final String key;
+  final bool isOptional;
+  Builder builder;
+  final List<String> requiredInputs;
+
+  TestBuilderDefinition(
+      this.key, this.isOptional, this.builder, this.requiredInputs);
+}
+
+/// Define a builder with key [key] that can be assembled into a `build.dart` or
+/// `build.yaml`.
+///
+/// [key] must match the top level variable name of [builder]. It must exist in
+/// the invoking script between the `test-package-start/end` comments. The
+/// builder should only use references to `package:build_test/build_test.dart`.
+///
+/// [requiredInputs] only has effect when using this builder with
+/// [packageWithBuilders]. In the [packageWithBuildScript] use case the ordering
+/// of the `builders` argument determines builder ordering.
+TestBuilderDefinition builder(String key, Builder builder,
+        {bool isOptional: false, List<String> requiredInputs: const []}) =>
+    new TestBuilderDefinition(key, isOptional, builder, requiredInputs);
+
+/// Create a package in [d.sandbox] with a `buid.yaml` file exporting [builders]
+/// and auto applying them to dependents.
+///
+/// The content in between `test-package-start/end` comments of the script that
+/// this function is called from will be copied into 'lib/builders.dart'. It
+/// should contain top level fields with names matching they keys in [builders]
+/// and only rely on imports to `package:build_test/build_test.dart`.
+d.Descriptor packageWithBuilders(Iterable<TestBuilderDefinition> builders,
+        {String name: 'provides_builders'}) =>
+    d.dir(name, [
+      _pubspec(name),
+      d.file('build.yaml', jsonEncode(_buildConfig(builders))),
+      d.dir('lib', [
+        d.file('builders.dart', _buildersFile(builders, new Frame.caller().uri))
+      ])
+    ]);
+
+Map _buildConfig(Iterable<TestBuilderDefinition> builders) => {
+      'builders':
+          builders.map(_builderDefinition).reduce((a, b) => a..addAll(b))
+    };
+
+Map _builderDefinition(TestBuilderDefinition builder) => {
+      builder.key: {
+        'import': 'package:provides_builders/builders.dart',
+        'builder_factories': ['${builder.key}Factory'],
+        'build_extensions': builder.builder.buildExtensions,
+        'auto_apply': 'dependents',
+        'is_optional': builder.isOptional,
+        'required_inputs': builder.requiredInputs,
+      }
+    };
+
+/// Create a package in [d.sandbox] with dependencies on [otherPackages] set up
+/// to run their builders with `pub run build_runner`.
+///
+/// The package name is always 'a'.
+///
+/// Files other than the pubspec should be set up with [packageContents].
+Future<BuildTool> package(Iterable<d.Descriptor> otherPackages,
+    {Iterable<d.Descriptor> packageContents}) async {
+  await d
+      .dir(
+          'a',
+          <d.Descriptor>[
+            await _pubspecWithDeps('a',
+                currentIsolateDependencies: ['build_runner', 'build'],
```

Comment:
should we add more dependencies here? if/when we override other dependencies by `path` this might break in the future (same below).

Specifically, build_config, build_resolvers, etc (the ones you usually get transitively)

Reply:
Done